File: packages/emitter/testcases/common/expected/types/function-type-aliases/GenericAliases.cs
namespace TestCases.common.types.functiontypealiases
{
    [global::Tsonic.Internal.ModuleContainerAttribute]
    public static class GenericAliases
    {
        public static bool test<T>(T value, global::System.Func<T, bool> pred)
        {
            return pred(value);
        }

        public static U transform<T, U>(T value, global::System.Func<T, U> fn)
        {
            return fn(value);
        }

        public static int compare<T>(T a, T b, global::System.Func<T, T, int> cmp)
        {
            return cmp(a, b);
        }
    }
}

File: packages/targets/csharp/emitter/testcases/common/types/function-type-aliases/GenericAliases.ts
import { int } from "@tsonic/core/types.js";

// Generic function type aliases
export type Predicate<T> = (x: T) => boolean;
export type Transform<T, U> = (x: T) => U;
export type Comparer<T> = (a: T, b: T) => int;

// Functions using type aliases
export function test<T>(value: T, pred: Predicate<T>): boolean {
  return pred(value);
}

export function transform<T, U>(value: T, fn: Transform<T, U>): U {
  return fn(value);
}

export function compare<T>(a: T, b: T, cmp: Comparer<T>): int {
  return cmp(a, b);
}
